### PR TITLE
Use `from __future__ import annotations`

### LINKED ISSUE #862: alternative constructors shouldn't hardcode their return types
Currently, alternative constructors like `Collection.from_dict` define the return type as `"Collection"`. https://github.com/stac-utils/pystac/blob/2e76a877b02eccf4d8e8ec6ad7b4aa7cf09aa268/pystac/collection.py#L607-L615

This makes is hard for subclasses to properly override the methods, since they'll be returning the subclasses type.

See something like https://stackoverflow.com/a/44644576/1889400 or https://mypy.readthedocs.io/en/stable/more_types.html?highlight=classmethod#precise-typing-of-alternative-constructors.

### PR DESCRIPTION
**Pull request by gadomski**
**Related Issue(s):**

- Supports #862

**Tl;dr:**

Switch to using `from __future__ import annotations` to give us slightly cleaner type signatures.

**Description:**

As described in the [mypy documentation](https://mypy.readthedocs.io/en/stable/runtime_troubles.html), there are three tools at our disposal when resolving runtime typing issues:

- `from __future__ import annotations`
- String types
- `typing.TYPE_CHECKING`

Currently, **pystac** uses string types and `typing.TYPE_CHECKING`. This works fine, but does lead to a bit of clutter in the typing namespace, e.g.: https://github.com/stac-utils/pystac/blob/f44de3ca86a8bc1cc1af19add4ac220fd0300365/pystac/asset.py#L8-L11

`"Item_Type"` and friends work fine for the purposes of defining our type signatures, but (IMO) they feel like a workaround and are a little awkward. Additionally, names like `"Item_Type"` will be confusing when combined with the names required to fix #862, e.g.:

```python
TItem = TypeVar("TItem", bound=Item)
if typing.TYPE_CHECKING:
    from pystac.item import Item as Item_Type

# We now have both `TItem` and `Item_Type` in scope, and that feels confusing

class Item:
    def from_dict(d: Dict[str, Any]) -> TItem:
        ....
```

This PR removes the `*_Type` names, and uses `from __future__ import annotations` throughout the library to simplify some type signatures. I don't believe this is a risky change, because (a) it's recommended by mypy as a potential solution to typing issues, and (b) this _may_ become the default Python behavior in the future, in which case we're setting ourselves up for success later.

There are no functional changes to behavior in this PR, this is purely a typing change. Also, because names like `Item_Type` are only used internally, I believe this PR is a non-breaking change.

As a sidecar fix, I also removed two unused imports in the top-level `__init__.py`.

**PR Checklist:**

- [x] Code is formatted (run `pre-commit run --all-files`)
- [x] Tests pass (run `scripts/test`)
- [ ] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/main/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
> There are no functional changes to behavior in this PR, this is purely a typing change. Also, because names like Item_Type are only used internally, I believe this PR is a non-breaking change.

If we're worried about this, we could leave stubs for these names around for a bit that emit a warning if they're accessed. It seems that pystac-client at least is OK: https://github.com/stac-utils/pystac-client/search?q=Item_Type. With a module-level `__getattr__` (https://peps.python.org/pep-0562/) this wouldn't be *too* much of a hassle.

**Comment by gadomski** [comment]
> If we're worried about this, we could leave stubs for these names around for a bit that emit a warning if they're accessed.

After playing some more, I think it's not worth worrying about. As a test, I tried to import `Item_Type` on current **main**, and got the following **mypy** error:

Screen Shot 2023-01-23 at 9 07 39 AM

So it appears there's decent guardrails in place already to ensure no one is using these "internal" type names.

**Review by duckontheweb — APPROVED** [review]
This looks really good to me; much cleaner. `__future__` to the rescue!

File: pystac/__init__.py [hunk 1]
```diff
@@ -41,7 +41,7 @@
 ]
 
 import os
-from typing import Any, AnyStr, Dict, Optional, Union
+from typing import Any, Dict, Optional
```
[the remark this answers is not in the document]
Comment (gadomski) [reply pystac/__init__.py:44]: `__init__.py` allows unused imports: https://github.com/stac-utils/pystac/blob/2d012251bda988d323939c28cba6f7a3b62e8898/.flake8#L4
